### ISSUE TITLE
Option for choosing forwarding mode

### ISSUE THREAD

**Issue by ghost** [TechnitiumSoftware/DnsServer#481]
Hello Shreyas,

this is not an issue, but a feature request.

Maybe it is possible to add an option to choose the behaviour of query forwarding.

If several forwarders are configured within Technitium DNS and if I am right, at the moment all of them are queried in parallel and the fastest server "wins". In my opinion, it would be nice to have also the option to query them serial (one by one) and, if none of the configured forwarders respond, to use instead recursion as a fallback (forward first mode) despite of the potential lack on privacy.

In my setup several Technitium DNS are forwarding to one private ISC BIND DoT test server. If the server does not respond for some reason, then the query fails at all. A serial query option will it also make possible, to forward a query to e. g. Cloudflare only if needed.

Sorry for my bad English. Last, I want to thank you very much for considering my suggestion and especially for this great and very helpful program. You made it easy (and possible) for me to get in an intensive touch with DoT.

Best regards,
Christian Jung

**Comment by ShreyasZare** [comment]
Thanks for the post. The idea that you have thought about seems to be about improving reliability from what you have described. But its not going to help in reality for too many reasons, some of which I am describing below.

The client devices are configured to timeout in 2 seconds. That means the DNS server has only 2 seconds to respond. The sequence of queries you have described is going to take too long to complete. Just increasing retries value in settings will help by keeping retrying until you get a response from the forwarding server. This approach will give response in least amount of time.

In 99.99% cases, the issue is not with the forwarding server. Most common reason usually is your own Internet connection causing the resolutions to fail. So, it does not matter if you switch to the next forwarder or start recursive resolution process.

Second most common reason is that the authoritative name servers for the domain name to be resolved are not responding quick enough. So, it still does not matter if you switch to the next forwarder or start recursive resolution process as the name servers for that domain are not going to suddenly start responding.

The current way that domains are resolved via forwarding servers uses concurrency to save time by using the fastest response that it gets. The default concurrency is set to 2 which means that 2 outstanding queries exists at any time if you have more forwarders configured.

Switching to recursive resolution when forwarders fail to respond is not going to work since if those forwarders are unable to resolve the domain name then so is the local DNS server going to fail to recursively resolve it.

**Comment by ChJungGH** [comment]
Hello Shreyas,

thank you very much for your detailed response.

You are absolutely right, in cases of network problems this makes not always sense. But I missed such a feature especially, when I do maintenance on my servers. For example, one of my Technitium DNS instances runs within Docker on my NAS. This server, which acts as a central cache, is configured as a forwarder on another machine. If I do system updates, the NAS is unreachable for about 10 minutes and the clients cannot resolve any addresses.

In this case it would be nice, if Technitium can automatically cache the unreachability for a given amount of time, skip this server and fallback to recursive resolution or switches to the next forwarder in the list (if there is another one configured). Of course, the clients may be unable to resolve some names, but, in my opinion, much less then without this fallback.

Maybe I am wrong with my thoughts, but it's my pleasure to discuss them with you.

Thanks again and best regards,
Christian Jung

**Comment by ShreyasZare** [comment]
> You are absolutely right, in cases of network problems this makes not always sense. But I missed such a feature especially, when I do maintenance on my servers. For example, one of my Technitium DNS instances runs within Docker on my NAS. This server, which acts as the primary resolver, is configured as a forwarder on another machine. If I do system updates, the NAS is unreachable for about 10 minutes and the clients cannot resolve any addresses.

In such case, its best that you run two instances of DNS servers and configure your clients to use both DNS servers so you have redundancy when any of the DNS server fails.

> In this case it would be nice, if Technitium can automatically cache the unreachability for a given amount of time, skip this server and fallback to recursive resolution or switches to the next forwarder in the list (if there is another one configured). Of course, the clients may be unable to resolve some names, but, in my opinion, much less then without this fallback.

The problem here is that the DNS server cannot know why a forwarder is not responding. It can be that the forwarder itself has some issue, or your internet has issue, or some random network delays/packet drops, etc. When it get a next query for resolution, the best way is to retry again to same forwarders instead of caching it as unreachable since if there is any network issue then it will cache all forwarders as unreachable. There are too many dynamic scenarios at play here so such a solution that you are proposing will work for only one specific scenario but will create too many issues when something else happens.

Switching to recursive resolution when forwarders fail is kind of very bad option. Recursive resolution takes a lot of time as it has to query at least 3 different name servers to get a response and if the response is a CNAME then it has to again do recursive resolution to resolve the CNAME domain. So, switching this way is just going to make it worse if there is an issue with network for few minutes.

For temporary issues, the DNS Cache with Serve Stale feature should be adequate where it keeps old stale records for reuse when network fails to resolve the domain name. So, if you keep the DNS server running 24x7 then it will have cached all the regular domain names that you query and even if you internet goes down, it will respond to queries from the stale cache data.

Best way is that you configure your DNS servers with a different strategy for the failure issues that you are planning.

> Thanks again and best regards, Christian Jung

You're welcome.
